### PR TITLE
Optional Parameters for ApexAnonymous Tasks

### PR DESCRIPTION
**Pull request by force2b**
# Critical Changes

# Changes

- Add support for up to two optional parameters in Apex Anonymous via token substitution.

# Issues Closed

### CODE REVIEW

File: cumulusci/tasks/apex/tests/test_apex_tasks.py [hunk 1]
```diff
@@ -614,6 +615,18 @@ def test_prepare_apex(self):
         expected = "String abc__str = 'foo';"
         self.assertEqual(expected, task._prepare_apex(before))
 
+    def test_optional_parameter_1_replacment(self):
```
Comment (davidmreed) [inline cumulusci/tasks/apex/tests/test_apex_tasks.py]:
```suggestion
    def test_optional_parameter_1_replacement(self):
```

File: cumulusci/tasks/apex/tests/test_apex_tasks.py [hunk 2]
```diff
@@ -614,6 +615,18 @@ def test_prepare_apex(self):
...
+        task = AnonymousApexTask(self.project_config, self.task_config, self.org_config)
+        before = "String str = '%%%PARAM_1%%%';"
+        expected = "String str = 'StringValue';"
+        self.assertEqual(expected, task._prepare_apex(before))
+
+    def test_optional_parameter_2_replacment(self):
```
Comment (davidmreed) [inline cumulusci/tasks/apex/tests/test_apex_tasks.py]:
```suggestion
    def test_optional_parameter_2_replacement(self):
```

File: docs/tasks.rst [hunk 3]
```diff
@@ -288,6 +288,8 @@ Options:
 * **apex**: A string of Apex to run (after the file, if specified).
 * **managed**: If True, will insert the project's namespace prefix.  Defaults to False or no namespace.
 * **namespaced**: If True, the tokens %%%NAMESPACED_RT%%% and %%%namespaced%%% will get replaced with the namespace prefix for Record Types.
+* **param1**: Optional parameter to pass to apex. Tokenized as %%%PARAM_1%%% in the Apex code
```
Comment (davidmreed) [inline docs/tasks.rst]:
```suggestion
* **param1**: Optional parameter to pass to Apex. Tokenized as %%%PARAM_1%%% in the Apex code.
```

File: docs/tasks.rst [hunk 4]
```diff
@@ -288,6 +288,8 @@ Options:
...
+* **param2**: Optional parameter to pass to apex. Tokenized as %%%PARAM_2%%% in the Apex code
```
Comment (davidmreed) [inline docs/tasks.rst]:
```suggestion
* **param2**: Optional parameter to pass to Apex. Tokenized as %%%PARAM_2%%% in the Apex code.
```

File: cumulusci/tasks/apex/anon.py [hunk 5]
```diff
@@ -35,6 +35,20 @@ class AnonymousApexTask(BaseSalesforceApiTask):
...
+        "param1": {
+            "description": (
+                "Parameter to pass to the apex. Use as %%%PARAM_1%%% in the Apex code"
+                "Defaults to an empty value."
+            ),
+            "required": False,
+        },
+        "param2": {
+            "description": (
+                "Parameter to pass to the apex. Use as %%%PARAM_2%%% in the Apex code"
```
Comment (davidmreed) [inline cumulusci/tasks/apex/anon.py]:
```suggestion
                "Parameter to pass to the Apex. Use as %%%PARAM_2%%% in the Apex code."
```

File: cumulusci/tasks/apex/anon.py [hunk 6]
```diff
@@ -35,6 +35,20 @@ class AnonymousApexTask(BaseSalesforceApiTask):
             ),
             "required": False,
         },
+        "param1": {
+            "description": (
+                "Parameter to pass to the apex. Use as %%%PARAM_1%%% in the Apex code"
```
Comment (davidmreed) [inline cumulusci/tasks/apex/anon.py]:
```suggestion
                "Parameter to pass to the Apex. Use as %%%PARAM_1%%% in the Apex code."
```

**Comment by force2b** [comment]
Interesting question @jstvz. Accepting an unlimited number of parameters would be a plus, though in terms of passing values to apex via the CLI is, 2 probably sufficient for the majority of use cases. I will say that key:value pairs as a command line parameter is probably better than what I implemented here, however I don't really have the python skills to rebuild this to use key:value pairs.

**Pull request merged**